File: backend/src/services/emailService.ts
import nodemailer from 'nodemailer';

export interface EmailConfig {
  host: string;
  port: number;
  secure: boolean;
  auth: {
    user: string;
    pass: string;
  };
}

export interface EmailOptions {
  to: string;
  subject: string;
  html: string;
  text?: string;
}

/**
 * Email service for sending authentication-related emails
 */
class EmailService {
  private transporter: nodemailer.Transporter | null = null;
  private isConfigured = false;

  constructor() {
    this.initializeTransporter();
  }

  /**
   * Initialize email transporter based on environment variables
   */
  private initializeTransporter() {
    const emailConfig = this.getEmailConfig();
    
    if (!emailConfig) {
      console.warn('Email service not configured. Password reset emails will not be sent.');
      return;
    }

    try {
      this.transporter = nodemailer.createTransport(emailConfig);
      this.isConfigured = true;
      console.log('✅ Email service configured successfully');
    } catch (error) {
      console.error('❌ Failed to configure email service:', error);
      this.isConfigured = false;
    }
  }

  /**
   * Get email configuration from environment variables
   */
  private getEmailConfig(): EmailConfig | null {
    const host = process.env.SMTP_HOST;
    const port = process.env.SMTP_PORT;
    const user = process.env.SMTP_USER;
    const pass = process.env.SMTP_PASS;

    if (!host || !port || !user || !pass) {
      return null;
    }

    return {
      host,
      port: parseInt(port, 10),
      secure: process.env.SMTP_SECURE === 'true',
      auth: {
        user,
        pass,
      },
    };
  }

  /**
   * Send email using configured transporter
   */
  async sendEmail(options: EmailOptions): Promise<boolean> {
    if (!this.isConfigured || !this.transporter) {
      console.error('Email service not configured');
      return false;
    }

    try {
      const info = await this.transporter.sendMail({
        from: process.env.SMTP_FROM || process.env.SMTP_USER,
        to: options.to,
        subject: options.subject,
        html: options.html,
        text: options.text,
      });

      console.log('📧 Email sent successfully:', info.messageId);
      return true;
    } catch (error) {
      console.error('❌ Failed to send email:', error);
      return false;
    }
  }

  /**
   * Send password reset email
   */
  async sendPasswordResetEmail(
    email: string,
    resetToken: string,
    firstName: string
  ): Promise<boolean> {
    const resetUrl = `${process.env.FRONTEND_URL || 'http://localhost:3000'}/reset-password?token=${resetToken}`;
    
    const html = `
      <!DOCTYPE html>
      <html>
        <head>
          <meta charset="utf-8">
          <meta name="viewport" content="width=device-width, initial-scale=1.0">
          <title>Password Reset</title>
          <style>
            body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
            .container { max-width: 600px; margin: 0 auto; padding: 20px; }
            .header { background: #6366f1; color: white; padding: 20px; text-align: center; border-radius: 8px 8px 0 0; }
            .content { background: #f9fafb; padding: 30px; border-radius: 0 0 8px 8px; }
            .button { display: inline-block; background: #6366f1; color: white; padding: 12px 24px; text-decoration: none; border-radius: 6px; margin: 20px 0; }
            .footer { text-align: center; margin-top: 30px; color: #6b7280; font-size: 14px; }
            .warning { background: #fef3c7; border: 1px solid #f59e0b; padding: 15px; border-radius: 6px; margin: 20px 0; }
          </style>
        </head>
        <body>
          <div class="container">
            <div class="header">
              <h1>Password Reset Request</h1>
            </div>
            <div class="content">
              <p>Hi ${firstName},</p>
              
              <p>We received a request to reset your password for your Project Management Platform account.</p>
              
              <p>Click the button below to reset your password:</p>
              
              <a href="${resetUrl}" class="button">Reset Password</a>
              
              <p>Or copy and paste this link into your browser:</p>
              <p style="word-break: break-all; background: #e5e7eb; padding: 10px; border-radius: 4px;">${resetUrl}</p>
              
              <div class="warning">
                <strong>Important:</strong> This link will expire in 1 hour for security reasons.
              </div>
              
              <p>If you didn't request this password reset, you can safely ignore this email. Your password will remain unchanged.</p>
              
              <p>Best regards,<br>The Project Management Platform Team</p>
            </div>
            <div class="footer">
              <p>This is an automated message. Please do not reply to this email.</p>
            </div>
          </div>
        </body>
      </html>
    `;

    const text = `
      Hi ${firstName},

      We received a request to reset your password for your Project Management Platform account.

      Please visit the following link to reset your password:
      ${resetUrl}

      This link will expire in 1 hour for security reasons.

      If you didn't request this password reset, you can safely ignore this email.

      Best regards,
      The Project Management Platform Team
    `;

    return this.sendEmail({
      to: email,
      subject: 'Reset Your Password - Project Management Platform',
      html,
      text,
    });
  }

  /**
   * Send welcome email to new users
   */
  async sendWelcomeEmail(email: string, firstName: string): Promise<boolean> {
    const loginUrl = `${process.env.FRONTEND_URL || 'http://localhost:3000'}/login`;
    
    const html = `
      <!DOCTYPE html>
      <html>
        <head>
          <meta charset="utf-8">
          <meta name="viewport" content="width=device-width, initial-scale=1.0">
          <title>Welcome to Project Management Platform</title>
          <style>
            body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
            .container { max-width: 600px; margin: 0 auto; padding: 20px; }
            .header { background: #6366f1; color: white; padding: 20px; text-align: center; border-radius: 8px 8px 0 0; }
            .content { background: #f9fafb; padding: 30px; border-radius: 0 0 8px 8px; }
            .button { display: inline-block; background: #6366f1; color: white; padding: 12px 24px; text-decoration: none; border-radius: 6px; margin: 20px 0; }
            .footer { text-align: center; margin-top: 30px; color: #6b7280; font-size: 14px; }
          </style>
        </head>
        <body>
          <div class="container">
            <div class="header">
              <h1>Welcome to Project Management Platform!</h1>
            </div>
            <div class="content">
              <p>Hi ${firstName},</p>
              
              <p>Welcome to Project Management Platform! Your account has been successfully created.</p>
              
              <p>You can now start organizing your projects, collaborating with your team, and tracking progress with our powerful project management tools.</p>
              
              <a href="${loginUrl}" class="button">Get Started</a>
              
              <p>If you have any questions or need help getting started, don't hesitate to reach out to our support team.</p>
              
              <p>Best regards,<br>The Project Management Platform Team</p>
            </div>
            <div class="footer">
              <p>This is an automated message. Please do not reply to this email.</p>
            </div>
          </div>
        </body>
      </html>
    `;

    const text = `
      Hi ${firstName},

      Welcome to Project Management Platform! Your account has been successfully created.

      You can now start organizing your projects, collaborating with your team, and tracking progress with our powerful project management tools.

      Visit ${loginUrl} to get started.

      If you have any questions or need help getting started, don't hesitate to reach out to our support team.

      Best regards,
      The Project Management Platform Team
    `;

    return this.sendEmail({
      to: email,
      subject: 'Welcome to Project Management Platform!',
      html,
      text,
    });
  }

  /**
   * Check if email service is configured and ready
   */
  isReady(): boolean {
    return this.isConfigured;
  }
}

// Export singleton instance
export const emailService = new EmailService();

File: internal/services/email_service.go
package services

import (
	"bytes"
	"fmt"
	"html/template"
	"log/slog"
	"time"

	"gopkg.in/gomail.v2"
)

// EmailService handles email operations
type EmailService interface {
	SendEmail(to, subject, body string) error
	SendHTMLEmail(to, subject, htmlBody, textBody string) error
	SendTemplateEmail(to, subject, templateName string, data interface{}) error
}

// SMTPEmailService implements EmailService using gomail with retry logic and error handling
type SMTPEmailService struct {
	host       string
	port       int
	username   string
	password   string
	from       string
	fromName   string
	templates  map[string]*template.Template
	logger     *slog.Logger
	maxRetries int
	retryDelay time.Duration
	secure     bool
}

// EmailConfig holds email configuration
type EmailConfig struct {
	SMTPHost     string
	SMTPPort     int
	SMTPUser     string
	SMTPPassword string
	FromAddress  string
	FromName     string
	Secure       bool
	MaxRetries   int
	RetryDelay   time.Duration
}

// NewEmailService creates a new SMTP email service with retry logic
func NewEmailService(config EmailConfig, logger *slog.Logger) EmailService {
	// Set defaults if not provided
	maxRetries := config.MaxRetries
	if maxRetries == 0 {
		maxRetries = 3
	}
	
	retryDelay := config.RetryDelay
	if retryDelay == 0 {
		retryDelay = 2 * time.Second
	}

	service := &SMTPEmailService{
		host:       config.SMTPHost,
		port:       config.SMTPPort,
		username:   config.SMTPUser,
		password:   config.SMTPPassword,
		from:       config.FromAddress,
		fromName:   config.FromName,
		templates:  make(map[string]*template.Template),
		logger:     logger,
		maxRetries: maxRetries,
		retryDelay: retryDelay,
		secure:     config.Secure,
	}

	// Load email templates
	service.loadTemplates()

	return service
}

// SendEmail sends a plain text email with retry logic
func (s *SMTPEmailService) SendEmail(to, subject, body string) error {
	// Skip sending email if SMTP is not configured
	if s.host == "" || s.host == "localhost" {
		s.logger.Info("Email would be sent (SMTP not configured)", "to", to, "subject", subject)
		return nil
	}

	m := gomail.NewMessage()
	m.SetHeader("From", fmt.Sprintf("%s <%s>", s.fromName, s.from))
	m.SetHeader("To", to)
	m.SetHeader("Subject", subject)
	m.SetBody("text/plain", body)

	return s.sendEmailWithRetry(m, to, subject)
}

// SendHTMLEmail sends an HTML email with optional text fallback and retry logic
func (s *SMTPEmailService) SendHTMLEmail(to, subject, htmlBody, textBody string) error {
	// Skip sending email if SMTP is not configured
	if s.host == "" || s.host == "localhost" {
		s.logger.Info("HTML Email would be sent (SMTP not configured)", "to", to, "subject", subject)
		return nil
	}

	m := gomail.NewMessage()
	m.SetHeader("From", fmt.Sprintf("%s <%s>", s.fromName, s.from))
	m.SetHeader("To", to)
	m.SetHeader("Subject", subject)
	
	if textBody != "" {
		m.SetBody("text/plain", textBody)
		m.AddAlternative("text/html", htmlBody)
	} else {
		m.SetBody("text/html", htmlBody)
	}

	return s.sendEmailWithRetry(m, to, subject)
}

// SendTemplateEmail sends an email using a template with retry logic
func (s *SMTPEmailService) SendTemplateEmail(to, subject, templateName string, data interface{}) error {
	tmpl, exists := s.templates[templateName]
	if !exists {
		s.logger.Error("Template not found", "template", templateName)
		return fmt.Errorf("template %s not found", templateName)
	}

	var htmlBuf, textBuf bytes.Buffer

	// Execute HTML template
	if err := tmpl.ExecuteTemplate(&htmlBuf, templateName+".html", data); err != nil {
		s.logger.Error("Failed to execute HTML template", "template", templateName, "error", err)
		return fmt.Errorf("failed to execute HTML template: %w", err)
	}

	// Execute text template (optional)
	textTemplate := templateName + ".txt"
	if tmpl.Lookup(textTemplate) != nil {
		if err := tmpl.ExecuteTemplate(&textBuf, textTemplate, data); err != nil {
			s.logger.Error("Failed to execute text template", "template", textTemplate, "error", err)
			return fmt.Errorf("failed to execute text template: %w", err)
		}
	}

	return s.SendHTMLEmail(to, subject, htmlBuf.String(), textBuf.String())
}

// sendEmailWithRetry sends the email using SMTP with gomail and retry logic
func (s *SMTPEmailService) sendEmailWithRetry(m *gomail.Message, to, subject string) error {
	d := gomail.NewDialer(s.host, s.port, s.username, s.password)
	
	// Configure TLS/SSL settings
	if s.secure {
		d.SSL = true
	}
	
	var lastErr error
	
	for attempt := 1; attempt <= s.maxRetries; attempt++ {
		s.logger.Info("Attempting to send email", 
			"to", to, 
			"subject", subject, 
			"attempt", attempt, 
			"max_retries", s.maxRetries)
		
		if err := d.DialAndSend(m); err != nil {
			lastErr = err
			s.logger.Warn("Failed to send email", 
				"to", to, 
				"subject", subject, 
				"attempt", attempt, 
				"error", err)
			
			// Don't retry on the last attempt
			if attempt < s.maxRetries {
				s.logger.Info("Retrying email send", 
					"delay", s.retryDelay, 
					"next_attempt", attempt+1)
				time.Sleep(s.retryDelay)
			}
		} else {
			s.logger.Info("Email sent successfully", 
				"to", to, 
				"subject", subject, 
				"attempt", attempt)
			return nil
		}
	}
	
	s.logger.Error("Failed to send email after all retries", 
		"to", to, 
		"subject", subject, 
		"max_retries", s.maxRetries, 
		"final_error", lastErr)
	
	return fmt.Errorf("failed to send email after %d attempts: %w", s.maxRetries, lastErr)
}

// loadTemplates loads email templates
func (s *SMTPEmailService) loadTemplates() {
	// Email verification template
	verificationHTML := `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Verify Your Email</title>
    <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #007bff; color: white; padding: 20px; text-align: center; }
        .content { padding: 20px; background: #f9f9f9; }
        .button { display: inline-block; padding: 12px 24px; background: #007bff; color: white; text-decoration: none; border-radius: 4px; }
        .footer { padding: 20px; text-align: center; color: #666; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Verify Your Email Address</h1>
        </div>
        <div class="content">
            <p>Hi {{.FirstName}},</p>
            <p>Thank you for registering with our project management platform! Please click the button below to verify your email address:</p>
            <p style="text-align: center;">
                <a href="{{.VerificationURL}}" class="button">Verify Email Address</a>
            </p>
            <p>Or copy and paste this link into your browser:</p>
            <p><a href="{{.VerificationURL}}">{{.VerificationURL}}</a></p>
            <p>This link will expire in 24 hours.</p>
            <p>If you didn't create an account, please ignore this email.</p>
        </div>
        <div class="footer">
            <p>Best regards,<br>Project Management Team</p>
        </div>
    </div>
</body>
</html>`

	verificationText := `
Hi {{.FirstName}},

Thank you for registering with our project management platform! Please click the link below to verify your email address:

{{.VerificationURL}}

This link will expire in 24 hours.

If you didn't create an account, please ignore this email.

Best regards,
Project Management Team`

	// Password reset template
	resetHTML := `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Reset Your Password</title>
    <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: #dc3545; color: white; padding: 20px; text-align: center; }
        .content { padding: 20px; background: #f9f9f9; }
        .button { display: inline-block; padding: 12px 24px; background: #dc3545; color: white; text-decoration: none; border-radius: 4px; }
        .footer { padding: 20px; text-align: center; color: #666; font-size: 12px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Reset Your Password</h1>
        </div>
        <div class="content">
            <p>Hi {{.FirstName}},</p>
            <p>You requested to reset your password. Click the button below to create a new password:</p>
            <p style="text-align: center;">
                <a href="{{.ResetURL}}" class="button">Reset Password</a>
            </p>
            <p>Or copy and paste this link into your browser:</p>
            <p><a href="{{.ResetURL}}">{{.ResetURL}}</a></p>
            <p>This link will expire in 1 hour.</p>
            <p>If you didn't request this, please ignore this email and your password will remain unchanged.</p>
        </div>
        <div class="footer">
            <p>Best regards,<br>Project Management Team</p>
        </div>
    </div>
</body>
</html>`

	resetText := `
Hi {{.FirstName}},

You requested to reset your password. Click the link below to create a new password:

{{.ResetURL}}

This link will expire in 1 hour.

If you didn't request this, please ignore this email and your password will remain unchanged.

Best regards,
Project Management Team`

	// Parse templates
	verificationTmpl := template.Must(template.New("verification.html").Parse(verificationHTML))
	template.Must(verificationTmpl.New("verification.txt").Parse(verificationText))

	resetTmpl := template.Must(template.New("reset.html").Parse(resetHTML))
	template.Must(resetTmpl.New("reset.txt").Parse(resetText))

	s.templates["verification"] = verificationTmpl
	s.templates["reset"] = resetTmpl
}

// MockEmailService is a mock implementation for testing
type MockEmailService struct {
	SentEmails []MockEmail
}

// MockEmail represents a sent email for testing
type MockEmail struct {
	To      string
	Subject string
	Body    string
	IsHTML  bool
}

// NewMockEmailService creates a new mock email service
func NewMockEmailService() *MockEmailService {
	return &MockEmailService{
		SentEmails: make([]MockEmail, 0),
	}
}

// SendEmail records the email instead of sending it
func (m *MockEmailService) SendEmail(to, subject, body string) error {
	m.SentEmails = append(m.SentEmails, MockEmail{
		To:      to,
		Subject: subject,
		Body:    body,
		IsHTML:  false,
	})
	return nil
}

// SendHTMLEmail records the HTML email instead of sending it
func (m *MockEmailService) SendHTMLEmail(to, subject, htmlBody, textBody string) error {
	m.SentEmails = append(m.SentEmails, MockEmail{
		To:      to,
		Subject: subject,
		Body:    htmlBody,
		IsHTML:  true,
	})
	return nil
}

// SendTemplateEmail records the template email instead of sending it
func (m *MockEmailService) SendTemplateEmail(to, subject, templateName string, data interface{}) error {
	m.SentEmails = append(m.SentEmails, MockEmail{
		To:      to,
		Subject: subject,
		Body:    fmt.Sprintf("Template: %s, Data: %+v", templateName, data),
		IsHTML:  true,
	})
	return nil
}

// GetLastEmail returns the last sent email
func (m *MockEmailService) GetLastEmail() *MockEmail {
	if len(m.SentEmails) == 0 {
		return nil
	}
	return &m.SentEmails[len(m.SentEmails)-1]
}

// Clear clears all sent emails
func (m *MockEmailService) Clear() {
	m.SentEmails = make([]MockEmail, 0)
}